FirebaseCore/Sources/FIRHeartbeatLogger.m

```diff
@@ -14,11 +14,30 @@
 
 #import <Foundation/Foundation.h>
 
-#import "FirebaseCore/Extension/FIRHeartbeatLogger.h"
-
 @import FirebaseCoreInternal;
 
 #import "FirebaseCore/Extension/FIRAppInternal.h"
+#import "FirebaseCore/Extension/FIRHeartbeatLogger.h"
+
+// MARK: - FIRWeakContainer
+
+/// A class used to weakly box reference types. This is used below to weakly store
+/// `FIRHeartbeatLogger` instances in a dictionary.
+@interface FIRWeakContainer<Object> : NSObject
+@property(readonly, weak) Object object;
+@end
+
+@implementation FIRWeakContainer
+
++ (instancetype)containerWithObject:(id)obj {
+  FIRWeakContainer *container = [[FIRWeakContainer alloc] init];
+  container->_object = obj;
```

Comment:
this is an interesting way to bypass readonly property and set the ivar value?

Reply:
Yep, no setter is generated because of `readonly` but you can bypass. this within the implementation by changing the underlying ivar.